### ISSUE TITLE
Starting playback extremely slow???

### ISSUE THREAD

**Issue by FCrane** [mpv-player/mpv#12753]
### Important Information

Provide following Information:
- mpv version: Nightly from 22.10.2023, Shinchiro build, libmpv-2.dll
- Windows Version: 11
- Source of the mpv binary: see above
- If known which version of mpv introduced the problem: not sure

### Reproduction steps
When starting playback, it sometimes takes extremely long (30 seconds or more) until the file plays. Starting it a second time usually is fast. The log indicates that there is some issue in libplacebo (complete log attached):

[   1.299][d][vo/gpu-next/libplacebo] [ 25] SPIRV_Cross_Output main(SPIRV_Cross_Input stage_input)
[   1.299][d][vo/gpu-next/libplacebo] [ 26] {
[   1.299][d][vo/gpu-next/libplacebo] [ 27]     _11 = stage_input._11;
[   1.299][d][vo/gpu-next/libplacebo] [ 28]     _15 = stage_input._15;
[   1.299][d][vo/gpu-next/libplacebo] [ 29]     vert_main();
[   1.299][d][vo/gpu-next/libplacebo] [ 30]     SPIRV_Cross_Output stage_output;
[   1.299][d][vo/gpu-next/libplacebo] [ 31]     stage_output.gl_Position = gl_Position;
[   1.299][d][vo/gpu-next/libplacebo] [ 32]     stage_output._9 = _9;
[   1.299][d][vo/gpu-next/libplacebo] [ 33]     return stage_output;
[   1.299][d][vo/gpu-next/libplacebo] [ 34] }
[   4.765][d][vo/gpu-next/libplacebo] shaderc compile status 'success' (0 errors, 0 warnings)
[   4.766][w][vo/gpu-next/libplacebo] Spent 3467.014 ms translating GLSL to SPIR-V (slow!)
[   4.927][v][vo/gpu-next/libplacebo] Spent 160.980 ms translating SPIR-V to HLSL (slow!)
[  44.470][w][vo/gpu-next/libplacebo] Spent 39542.673 ms translating HLSL to DXBC (slow!)
[  44.470][d][vo/gpu-next/libplacebo] fragment shader HLSL source:
[  44.470][d][vo/gpu-next/libplacebo] [  1] cbuffer _25_27 : register(b0)
[  44.470][d][vo/gpu-next/libplacebo] [  2] {
[  44.470][d][vo/gpu-next/libplacebo] [  3]     float2 _27_m0 : packoffset(c0);
[  44.470][d][vo/gpu-next/libplacebo] [  4]     row_major float3x3 _27_m1 : packoffset(c1);
[  44.470][d][vo/gpu-next/libplacebo] [  5]     row_major float3x3 _27_m2 : packoffset(c4);
[  44.470][d][vo/gpu-next/libplacebo] [  6]     row_major float3x3 _27_m3 : packoffset(c7);
[  44.470][d][vo/gpu-next/libplacebo] [  7]     float _27_m4 : packoffset(c10);
[  44.470][d][vo/gpu-next/libplacebo] [  8]     float _27_m5 : packoffset(c10.y);
[  44.470][d][vo/gpu-next/libplacebo] [  9]     row_major float3x3 _27_m6 : packoffset(c11);
[  44.470][d][vo/gpu-next/libplacebo] [ 10] };

Something with the gsls shaders seems to be wrong... Didn't happen in previous versions and doesn't happen when starting the exact same file again...

### Expected behavior
Fast startup

### Actual behavior
Slow startup

### Log file

[log.zip](https://github.com/mpv-player/mpv/files/13178516/log.zip)

**Comment by llyyr** [comment]
Provide a log file with `--no-config` and the minimum required options set please.

**Comment by FCrane** [comment]
Here are the logs with no-config=yes. Included is the mpv.conf file and the log for the first start of an mkv (extremely slow) and the second start directly afterwards (fast).
[Logs noconfig.zip](https://github.com/mpv-player/mpv/files/13179875/Logs.noconfig.zip)

**Comment by llyyr** [comment]
`no-config=yes` isn't a valid option, you have to use `--no-config` as a command line flag for mpv.

**Comment by FCrane** [comment]
As I mentioned, I'm using the DLL version, so I can't pass the command line flag...

**Comment by FCrane** [comment]
Ok, I've used mpv_set_parameter in my front end to disable the config and then it starts fast. However, that does not help, because I need my config (of course).

I've now checked using vo = gpu and that also starts fast - so it seems to narrow down to gpu-next.

**Comment by FCrane** [comment]
Also strange: if I display the mpv video window before using the mpv command "loadfile", it also starts fast! My regular order ov things is to "loadfile", then get the video parameters and based on that resize the video window and show it. This has always worked fine, but recently it started to become slow. Why?

I prefer not to show a default player window (small size) before the video starts, of course, because this is ugly. I'd like mpv to start playing as fast as it always did...

**Comment by FCrane** [comment]
As you can see in my logfiles, libplacebo hangs extremely long when converting the glsl - why is that? Why didn't  it happen before? Why does it work fast when starting the same file a second time? And why does it still happen if I don't use any glsl shaders at all?

**Comment by haasn** [comment]
>` [ 44.470][w][vo/gpu-next/libplacebo] Spent 39542.673 ms translating HLSL to DXBC (slow!)`

This is handled entirely by closed-source d3dcompiler.dll. Good luck debugging it.

**Comment by FCrane** [comment]
With gpu-api=vulkan there's a different problem: when my program calls mpv_command "loadfile", the MPV_EVENT_VIDEO_RECONFIG arrives, but mpv_get_parameter "video-params/dw" returns MPV_ERROR_PROPERTY_UNAVAILABLE.

Why??? This works fine without vulkan and vulkan is supported (because when I start the same file again, it works!) How can the property be unavailable when loadfile starts the playback and there is a video stream present?

**Comment by kasper93** [comment]
> This is handled entirely by closed-source d3dcompiler.dll. Good luck debugging it.

The d3dcompiler indeed may be slow, but the main problem is that we feed huge shaders, mostly because of fully unrolling loops.

Those already takes unreasonable amount of time.
```
[ 4.766][w][vo/gpu-next/libplacebo] Spent 3467.014 ms translating GLSL to SPIR-V (slow!)
[ 4.927][v][vo/gpu-next/libplacebo] Spent 160.980 ms translating SPIR-V to HLSL (slow!)
```

**Comment by FCrane** [comment]
Any news on this issue?  Currently, all new versions of libmpv are useless to me because they start so extremely slowly...

**Comment by FCrane** [comment]
What does this issue have to do with downscaling?

**Comment by FCrane** [comment]
Any news on this issue? I just tried the latest Nightly of libmpv and the issue is still there! Startup is EXTREMELY slow... Older versions (about a year ago) work just fine, but I'd like to update of course to use all new features, fixes, etc.

Thanks!

**Comment by kasper93** [comment]
>I just tried the latest Nightly of libmpv and the issue is still there! Startup is EXTREMELY slow... 
Post config and log when the issue happens.
>Older versions (about a year ago) work just fine
Than it is very easy fix, go to https://sourceforge.net/projects/mpv-player-windows/files/64bit/ and find last working build, so we know where the regression happened.

**Comment by FCrane** [comment]
I did post logs above (even with --no-config). I'm using libmpv-2.dll so I don't really know when the issue started. Also, it's extremely hard to test different versions, because as I explained the issue only happens on first start of a media file - when I play the exact same file again, it starts fast. I need to restart the computer to make it slow again for the first start... Very strange!

Also, I observed the following: after a file started slowly, some others will start fast even the first time they are played. Seems that something in the d3d compiler "has already done its thing" with the first file and now does not need to do again with others that "need the same thing done", but other files need other things - sorry for the un-technical description. Maybe something gets cached, etc. until restart?

**Comment by kasper93** [comment]
> I did post logs above (even with --no-config).

The logs you posted are from version that had cache trashing issue:
```
Object 0xa66efcc8d4e8d607 (size 3145728) exceeds max size 1048576, discarding
```
so it is expected that it will "randomly" work slower. As the cache was not really effective, because it was clearing itself due to very small limit. It has been recently fixed and shouldn't be an issue anymore.

> Also, I observed the following: after a file started slowly, some others will start fast even the first time they are played. Seems that something in the d3d compiler "has already done its thing" with the first file and now does not need to do again with others that "need the same thing done", but other files need other things - sorry for the un-technical description. Maybe something gets cached, etc. until restart?

Indeed things are cached. In recent version, after 69891c4070f03236c3843fd49e6ee69a46dfe82c it no longer will "lose" old cached entries, so you see the slowdown only once. I suspect you are not using new version from your description.

Could you retest with the cache fix included? And if there are still issues, let me know when they happen.

**Comment by FCrane** [comment]
I'm using the nightly built of 04.02.2024, so it's almost the newest available...

I've noticed that the old version created a file "libplacebo.cache", but the new versions don't? Where is the cache saved now? What exactly is cached?

Also, the new versions have a 20-30% higher GPU load on my AMD RX550 (and also my MiniPC with a Ryzen 7 5800H integrated GPU), resulting in heavy stutter for some full res UHD videos :-( The old version sits around 80-90% (full screen UHD, 1 gsls shader for slight sharpening) and the new one maxes out at 100% almost all the time, producing stutter.

**Comment by kasper93** [comment]
> I'm using the nightly built of 04.02.2024, so it's almost the newest available...

Good, so you can post log from this version, so we see whole picture.

> I've noticed that the old version created a file "libplacebo.cache", but the new versions don't? Where is the cache saved now? What exactly is cached?

 It should produce separate cache files in the same directory as before. You should notice them.

> Also, the new versions have a 20-30% higher GPU load on my AMD RX550 (and also my MiniPC with a Ryzen 7 5800H integrated GPU), resulting in heavy stutter for some full res UHD videos :-(

I don't know about that, what was your previous version? There were no changes recently in rendering area.

**Comment by FCrane** [comment]
Here's an example. Although not extremely slow, still much too slow...

[log.zip](https://github.com/mpv-player/mpv/files/14186585/log.zip)

[mpv.zip](https://github.com/mpv-player/mpv/files/14186591/mpv.zip)

Seems that lines like these cause the long delays:
[   2.444][w][vo/gpu-next/libplacebo] Spent 1567.325 ms translating GLSL to SPIR-V (slow!)
[   2.516][v][vo/gpu-next/libplacebo] Spent 71.779 ms translating SPIR-V to HLSL
[   6.378][w][vo/gpu-next/libplacebo] Spent 3861.843 ms translating HLSL to DXBC (slow!)

These are the same kinds of issues like in my log at the beginning of the thread, so the problem always seems to have the same cause.

The new version definitely does NOT create a cache file at the same location as the exe. So where is it now? And what is in it that makes starting the same (or similar) file fast?

**Comment by kasper93** [comment]
> The new version definitely does NOT create a cache file at the same location as the exe. So where is it now?

[   0.222][d][global] cache path: '' -> 'C:\Users\Martin\AppData\Local/mpv/cache'

> And what is in it that makes starting the same (or similar) file fast?

Compiled shaders, those messages about `translating` should be gone and instead those should be loaded from cache. You should see `cache_load_obj` messages. Shaders can differ between files and window scales, but they should be cached eventually. 

> Here's an example. Although not extremely slow, still much too slow...

Can you try few times and let the cache populate itself? I see there are still uncompiled shaders being saved in your log. If after few times it still happes, please send another log and I can see if there is a real issue in something not being cached.

**Comment by FCrane** [comment]
Here are 2 more logs - delays of 10 seconds or more are annoying...

[log.zip](https://github.com/mpv-player/mpv/files/14194788/log.zip)
[log2.zip](https://github.com/mpv-player/mpv/files/14194790/log2.zip)

Which shaders are still "uncompiled"? Why can't mpv "precompile" all possibly needed shaders once (e.g. with a parameter "-compileshaders" that only needs to be started once)?

AND: why did previous versions not show these delays??? They obviously also built a shader cache but it was fast. Why can't this work as before? Shaders are now working as with previous versiosn, there does not seem to be any difference...

**Comment by kasper93** [comment]
> Which shaders are still "uncompiled"?

**Comment by kasper93** [comment]
> Which shaders are still "uncompiled"?

Scalers mostly.

> Why can't mpv "precompile" all possibly needed shaders once (e.g. with a parameter "-compileshaders" that only needs to be started once)?

Because it is not feasible to generate all shader permutations. Also once it is cached it is cached, so the complation time is spread around multiple runs. It is easier this way too.

> why did previous versions not show these delays???

I don't know what "previous version" you referring to, so I cannot be sure. I suspect that you are referring to version before shaders generation refactor in libplacebo. Now loops are fully unrolled and those loops can be big for scalers, which make huge shaders, that take long to process.

> They obviously also built a shader cache but it was fast. Why can't this work as before?

Like said above libplacebo unrolls loops agressively, which makes shaders slow to compile, cache is here to mitigate that, but you need to populate the cache first.

> Here are 2 more logs - delays of 10 seconds or more are annoying...

Is this happening all the time, if you run a file back-to-back or you are just finding new cases?

**Comment by FCrane** [comment]
> I don't know what "previous version" you referring to, so I cannot be sure. I suspect that you are referring to version before shaders generation refactor in libplacebo. Now loops are fully unrolled and those loops can be big for scalers, which make huge shaders, that take long to process.
> 
> > They obviously also built a shader cache but it was fast. Why can't this work as before?
> 
> Like said above libplacebo unrolls loops agressively, which makes shaders slow to compile, cache is here to mitigate that, but you need to populate the cache first.

The last "good" version I have is from 30.04.2023 (libmpv-2.dll). It does create a libplacebo.cache in the same folder as the exe, but it NEVER had any long delays at startup...

> Is this happening all the time, if you run a file back-to-back or you are just finding new cases?
It seems to happen once for (some) new files. Subsequent starts of the same file is fast. However, there always seem to be new "problematic" files - if it's really the scalers, it would mean that for every "new" video resolution this would happen? Very annoying, because videos are cropped to a million different resolutions...

Again: until some point (which I unfortunately cannot say when exactly, because I didn't upgrade the mpv dll every week or even month), this problem was non-existent! Start was always fast, for any new and old video. And all shaders, scalers, etc. worked perfectly, a libplacebo.cache was created (I actually prefer it to be in the same folder as the exe, so that I can backup it if I reinstall the OS, move to another computer, etc.). And the GPU load was 20-30% lower!

Why couldn't this be kept the perfect way it was? Can I skip the complete unrolling of loops with a parameter? It's really annoying, because it can take up to 30 seconds or more...

**Comment by Hrxn** [comment]
Do you have a virus scanner running or something?

**Comment by FCrane** [comment]
No virus scanner, dedicated HTPC.

Unfortunately, when testing today I found out, that the new dll also has horrible lip-sync issues! Lip-Sync is often way off, sometimes better, sometimes worse within the same file. Sometimes it can't even be corrected with setting the audio delay (on the fly), etc. Horrible and unusable! I'll open a new issue for that.

It's really a shame that mpv is getting worse and worse :-(

I need to stick with the version of 30.04.2023 forever as it is the last one I have that fully works...

**Comment by kasper93** [comment]
> It's really a shame that mpv is getting worse and worse :-(

Sorry that we failed you.

**Comment by FCrane** [comment]
Here's another example: the delay was more than 30 seconds!
[log.zip](https://github.com/mpv-player/mpv/files/14202677/log.zip)
The video was 3840x2160, which is the standard UHD and I've played many of them before on that installation, so it must have been already cached! It was not a "new" resolution, so there must be something else causing this...

**Comment by FCrane** [comment]
> > It's really a shame that mpv is getting worse and worse :-(
> 
> Sorry that we failed you.

Yes, it's apity! mpv used to be the best by far and I've even programmed my own front-end to use the dll. Worked great, superior to everything else! I still will use it, of course, but I'd really like to keep updated (for support of new video formats, HDR, DV, sound formats, etc.) but unfortunately, updates only make it slower, use more GPU, cause all sorts of video issues, audio issues, etc....

I really hope the team manages to make a "good" version again without all the new issues.

**Comment by kasper93** [comment]
> And the GPU load was 20-30% lower!

For this you should use `--profile=fast`

> The video was 3840x2160, which is the standard UHD and I've played many of them before on that installation, so it must have been already cached! It was not a "new" resolution, so there must be something else causing this...

It is a ratio between source and destination size, so it can happen multiple times. But "after a while" all cases should be cached. Anyway I've send PR to not unroll ortho kernel https://code.videolan.org/videolan/libplacebo/-/merge_requests/633 which will resolve default configuration issues.

**Comment by FCrane** [comment]
> For this you should use `--profile=fast`
But I need my profile settings for proper playback? Can I combine the profiles? What are the disadvantages of "fast"?

> It is a ratio between source and destination size, so it can happen multiple times.
Output size is always 3840x2160 (UHD projector), so maybe there is something that causes the shaders for same resolution videos to be recompiled?

**Comment by kasper93** [comment]
@FCrane: test this build https://github.com/zhongfly/mpv-winbuild/releases/tag/2024-02-08-a45518c

**Comment by FCrane** [comment]
Yes, this version works!!! Lip-sync is correct, starts super fast! THANK YOU!!!!

**Comment by kasper93** [comment]
Fixed by: https://github.com/haasn/libplacebo/commit/b2b61a755921499e2075d409a9eb667232be194b

**Comment by Hrxn** [comment]
Yay, so this can be closed then?
